File: samples/snippets/csharp/VS_Snippets_CFX/logarchievesnapshot/cs/logarchievesnapshot.cs
using System;
using System.IO.Log;
using System.IO;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace Snippets
{
    //<Snippet0>
    class LogBackup
    { 
        static void ArchiveToXML(LogStore logStore, string fileName) 
        {  
            LogArchiveSnapshot snapshot = logStore.CreateLogArchiveSnapshot();
            
            XmlTextWriter writer = new XmlTextWriter(fileName, Encoding.ASCII);
            
            writer.WriteStartElement("logArchive");
            foreach(FileRegion region in snapshot.ArchiveRegions) 
            {
                writer.WriteStartElement("fileRegion"); 
                writer.WriteElementString("path", region.Path); 
                writer.WriteElementString("length", region.FileLength.ToString()); 
                writer.WriteElementString("offset", region.Offset.ToString()); 
                using(Stream dataStream = region.GetStream()) 
                { 
                    byte[] data = new byte[dataStream.Length]; 
                    dataStream.Read(data, 0, data.Length); 
                    writer.WriteElementString("data", Convert.ToBase64String(data)); 
                }
                writer.WriteEndElement();
            }
            writer.WriteEndElement(); 
            writer.Close(); 
            logStore.SetArchiveTail(snapshot.LastSequenceNumber); 
            
        }
        static void RestoreFromXML(string fileName) 
        { 
            using(XmlTextReader reader = new XmlTextReader(fileName)) 
            {
                reader.ReadStartElement("logArchive"); 
                while(reader.IsStartElement()) 
                { 
                    string path = reader.ReadElementString("path"); 
                    long length = Int64.Parse(reader.ReadElementString("length")); 
                    long offset = Int64.Parse(reader.ReadElementString("offset")); 
                    string dataString = reader.ReadElementString("data"); 
                    byte[] data = Convert.FromBase64String(dataString); 
                    FileStream fileStream;
                    using(fileStream = new FileStream(path, FileMode.OpenOrCreate, FileAccess.Write)) 
                    { 
                        fileStream.SetLength(length); 
                        fileStream.Position = offset; fileStream.Write(data, 0, data.Length); 
                    }
                } 
            reader.ReadEndElement(); 
            } 
        } 
    }
    //</Snippet0>
}

File: samples/snippets/visualbasic/VS_Snippets_CFX/logarchievesnapshot/vb/logarchievesnapshot.vb
Imports System
Imports System.IO.Log
Imports System.IO
Imports System.Collections.Generic
Imports System.Text
Imports System.Xml

Namespace Snippets
    '<Snippet0>
    Friend Class LogBackup
        Private Shared Sub ArchiveToXML(ByVal logStore As LogStore, ByVal fileName As String)
            Dim snapshot As LogArchiveSnapshot = logStore.CreateLogArchiveSnapshot()

            Dim writer As New XmlTextWriter(fileName, Encoding.ASCII)

            writer.WriteStartElement("logArchive")
            For Each region As FileRegion In snapshot.ArchiveRegions
                writer.WriteStartElement("fileRegion")
                writer.WriteElementString("path", region.Path)
                writer.WriteElementString("length", region.FileLength.ToString())
                writer.WriteElementString("offset", region.Offset.ToString())
                Using dataStream As Stream = region.GetStream()
                    Dim data(dataStream.Length - 1) As Byte
                    dataStream.Read(data, 0, data.Length)
                    writer.WriteElementString("data", Convert.ToBase64String(data))
                End Using
                writer.WriteEndElement()
            Next region
            writer.WriteEndElement()
            writer.Close()
            logStore.SetArchiveTail(snapshot.LastSequenceNumber)

        End Sub
        Private Shared Sub RestoreFromXML(ByVal fileName As String)
            Using reader As New XmlTextReader(fileName)
                reader.ReadStartElement("logArchive")
                Do While reader.IsStartElement()
                    Dim path = reader.ReadElementString("path")
                    Dim length = Int64.Parse(reader.ReadElementString("length"))
                    Dim offset = Int64.Parse(reader.ReadElementString("offset"))
                    Dim dataString = reader.ReadElementString("data")
                    Dim data() = Convert.FromBase64String(dataString)
                    Dim fileStream As FileStream
                    fileStream = New FileStream(path, FileMode.OpenOrCreate, FileAccess.Write)
                    Using fileStream
                        fileStream.SetLength(length)
                        fileStream.Position = offset
                        fileStream.Write(data, 0, data.Length)
                    End Using
                Loop
                reader.ReadEndElement()
            End Using
        End Sub
    End Class
    '</Snippet0>
End Namespace